lib/spack/spack/concretize.py

```diff
@@ -203,46 +202,91 @@ def concretize_version(self, spec):
 
         return True   # Things changed
 
+    def _concretize_arch_component(self, platform, spec):
+        """ Concretizes the specified component of the architecture in a
+        spec.
+
+        On heterogeneous architectures it will look and see if the first
+        component will match the second component ie if the os is for the
+        front end then it will be sure to match the target with the front end.
+        On architectures with one single architecture it should work the same
+        since both front end and back end are equivalent.
+
+        Args:
+            component (str):      which component to check and set in arch spec
+            platform  (Platform): platform object to help set the attributes of
+                                  architecture 
+            spec      (Spec):     spec object to set the architecture on.
+
+        Returns:
+            Returns True if any of the components were set. If all components
+            are present then return False. True represents a change happened
+            and False represents no change."""
+        architecture = spec.architecture
+        #  represents whatever frontend exists, same if on single arch
+
+        #  methods available to the platform class depends on component
+        #  so created a mapping between component and method name.
+
+        if architecture.concrete:
+            return False
+
+        frontend_os = str(platform.operating_system("fe"))
+        frontend_target = str(platform.target("fe"))
+        default_os = str(platform.operating_system("default_os"))
```

Comment:
`default_os` is always either the backend or frontend. For what this function intends, I think it would be clearer to explicitly retrieve the backend os (e.g. with `"be"`)